src/style.rs

```diff
@@ -674,12 +687,99 @@ enum Alignment {
     Right,
 }
 
+/// Trait for defining stateful or stateless formatters
+///
+/// Stateless formatters are the more simple type and are implemented as a closure in the form of [`Fn(&ProgressState) -> String`]
+/// Stateful formatters can be anything that implements this trait
+pub trait ProgressTracker: Send {
+    fn default(&self) -> Box<dyn ProgressTracker>;
+    fn tick(&mut self, state: &ProgressState, now: Instant);
+    fn reset(&mut self, state: &ProgressState, now: Instant);
+    fn message(&self, state: &ProgressState) -> String;
```

Comment:
If you're still open to working on this, we should change `fn message(&self, state: &ProgressState) -> String` to `fn write(&self, state: &ProgressState, w: &mut dyn fmt::Write) -> fmt::Result`.